src/tls_builtin.c

```diff
@@ -623,7 +623,7 @@ static void mg_tls_server_send_hello(struct mg_connection *c) {
   // calculate keyshare
   uint8_t x25519_pub[X25519_BYTES];
   uint8_t x25519_prv[X25519_BYTES];
-  mg_random(x25519_prv, sizeof(x25519_prv));
+  if (!mg_random(x25519_prv, sizeof(x25519_prv))) mg_error(c, "RNG"); 
```

Comment:
Can't we just fail in mg_tls_init() instead of failing in three different places ? We can ignore result here and try a bogus random number on init; if it fails, we error (disconnect).

Reply:
Why should we ignore RNG failure after TLS init ?

Reply:
To avoid having the same check in 3 different places, and waiting for the Hello to fail; when we can just elegantly fail on connection establishment, which happens before any exchange, when the user calls mg_tls_init().
So, instead of forcing the TLS handshake to start, and then (in 3 different places) fail because we don't have a good RNG (which to me is read as a lack of prevision, no look ahead), we check that before attempting anything, we know we won't be able to fulfill the request, hence we fail right from the start.

Reply:
If we failed on TLS init, no TLS init has taken place, hence those calls can't happen, connection has been dropped off.

Reply:
To avoid having the same check in 3 different places, and waiting for the Hello to fail; when we can just elegantly fail on connection establishment, which happens before any exchange, when the user calls mg_tls_init().
So, instead of forcing the TLS handshake to start, and then (in 3 different places) fail because we don't have a good RNG (which to me is read as a lack of prevision, no look ahead), we check that before attempting anything, we know we won't be able to fulfill the request, hence we fail right from the start.
If init fails, the other calls don't have a chance to be made. Or do they ?

Reply:
My understanding is that mg_random() can succeed at one invocation, and fail in the other. That could be true for UNIX for sure - mg_random opens and reads /dev/urandom, which can succeed or fail depending on the state of the process. Thus I think we MUST check every single mg_random() invocation.

Reply:
I was convinced that the "weak" condition could be determined in one call.
So does this mean that some connections will work and other wills just fail ? That is a poor user experience. Is there any way to detect / remember the PRNG is a weak one and fail TLS attempts ?

Reply:
Yeah, TLS connection might unexpectedly fails, if RNG degrades. That's what the security researcher suggests. I guess we should follow the advice.

Reply:
Oh, I didn't know about degradation. Cool.